### PR TITLE
Consume dependency attributes

### PR DESCRIPTION
**Pull request by melix**
### Context

This pull request builds on top of #5006 and adds attributes to Gradle module metadata. It makes it possible to consume published modules which define attributes on dependencies.

Closes gradle/gradle-private#1190

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java [hunk 1]
```diff
@@ -16,78 +16,37 @@
...
-import java.util.Collection;
 import java.util.List;
-import java.util.Set;
 
 public class RejectedModuleMessageBuilder {
     public String buildFailureMessage(ModuleResolveState module) {
         boolean hasRejectAll = false;
         for (SelectorState candidate : module.getSelectors()) {
-            hasRejectAll |= isRejectAll(candidate.getVersionConstraint());
+            hasRejectAll |= MessageBuilderHelper.isRejectAll(candidate.getVersionConstraint());
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java]: This doesn't seem to belong on this helper, but on the class itself.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java [hunk 2]
```diff
@@ -16,78 +16,37 @@
...
         }
         StringBuilder sb = new StringBuilder();
         if (hasRejectAll) {
             sb.append("Module '").append(module.getId()).append("' has been rejected:\n");
         } else {
             sb.append("Cannot find a version of '").append(module.getId()).append("' that satisfies the version constraints: \n");
         }
-        for (EdgeState incomingEdge : getIncomingEdges(module)) {
+        for (EdgeState incomingEdge : MessageBuilderHelper.getIncomingEdges(module)) {
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java]: This doesn't seem to belong on this helper, but on the class itself.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java (hunk 1 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java]: It's shared code between 2 different error message builders (that's why it's in the helper).

Comment (oehme) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java]: I mean it probably belongs on the constraint class.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java (hunk 2 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java]: done

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java (hunk 1 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/RejectedModuleMessageBuilder.java]: Oh I see, done!

**Pull request merged**